processor/src/main/java/com/linecorp/decaton/processor/runtime/internal/PartitionContexts.java

```diff
@@ -245,28 +277,37 @@ public void partitionsResumed(List<TopicPartition> partitions) {
     }
 
     /**
-     * Waits for all pending tasks if property-reload is requested, then recreate all partition contexts with latest property values.
+     * Waits for pending tasks if property-reload is requested,
+     * then recreate partition contexts with latest property values.
      * This method must be called from only subscription thread.
      */
     public void maybeHandlePropertyReload() {
-        if (reloadRequested.get()) {
-            if (totalPendingTasks() > 0) {
-                logger.debug("Waiting pending tasks for property reload.");
+        propertyReloadLock.lock();
+        try {
+            List<TopicPartition> reloadableTopicPartitions = contexts.entrySet()
+                    .stream()
+                    .filter(entry -> entry.getValue().reloadRequested()
+                                     && entry.getValue().pendingTasksCount() == 0)
+                    .map(Entry::getKey)
+                    .collect(Collectors.toList());
+            if (reloadableTopicPartitions.isEmpty()) {
                 return;
             }
-            // it's ok to check-and-set reloadRequested without synchronization
-            // because this field is set to false only in this method, and this method is called from only subscription thread.
-            reloadRequested.set(false);
-            logger.info("Completed waiting pending tasks. Start reloading partition contexts");
-            reloadContexts();
+            reloadContexts(reloadableTopicPartitions);
+            long reloadingPartitions = contexts.values()
+                    .stream()
+                    .filter(PartitionContext::reloadRequested)
+                    .count();
+            if (reloadingPartitions == 0) {
+                logger.info("Completed reloading all partition contexts");
+            }
+        } finally {
+            propertyReloadLock.unlock();
         }
     }
 
-    private void reloadContexts() {
-        // Save current topicPartitions into copy to update contexts map while iterating over this copy.
-        Set<TopicPartition> topicPartitions = new HashSet<>(contexts.keySet());
-
-        logger.info("Start dropping partition contexts");
+    private void reloadContexts(Collection<TopicPartition> topicPartitions) {
+        logger.info("Start dropping partition context({})", topicPartitions);
         removePartition(topicPartitions);
         logger.info("Finished dropping partition contexts. Start recreating partition contexts");
```

Comment:
[nits] This log and this line https://github.com/line/decaton/blob/5659ec7019e70f857f7d39aa86b301fdb69388e5/processor/src/main/java/com/linecorp/decaton/processor/runtime/internal/PartitionContexts.java#L316 should also include `topicPartitions`